userguide/tutorials/plugin_development.adoc

```diff
@@ -189,6 +189,57 @@ When you start developing your own plugin, it would be useful to be able to set
 
 . This runs the application in debug mode. You can also set additional breakpoints as required.
 
+=== Authentication within plugins
+
+In order to invoke write API operations like `AccountUserApi#createAccount`, plugin code must authenticate against Kill Bill first. Otherwise, it will result in an `org.apache.shiro.authz.UnauthenticatedException`. This section explains how authentication can be done.
+
+==== Authentication steps
+
+In order to authenticate against Kill Bill, the following needs to be done:
+
+1. Obtain authentication credentials - Although plugins can use the `admin/password` credentials for authentication, it is typically not advisable to do so. This is to limit the scope of operations that plugins can do. It is thus recommended to https://killbill.github.io/slate/#tenant-add-a-per-tenant-configuration-for-a-plugin[configure plugins with custom credentials] and retrieve them in the code. You can see an example of retrieving custom credentials in the https://github.com/killbill/killbill-email-notifications-plugin/blob/8df47156a2e80c65ce574e0ad689afd02b926f59/src/main/java/org/killbill/billing/plugin/notification/setup/EmailNotificationListener.java#L230[Email Notification Listener] code.
+
+2. Add the following code before invoking any API operation (use credentials obtained above):
+
+[source,java]
+killbillAPI.getSecurityApi().login(login, password);
+
+[start=3]
+3. Invoke the necessary api method:
+
+[source,java]
+invoiceUserApi.triggerDryRunInvoiceGeneration(accountId, targetDate, dryRunArguments, callContext);
+
+[start=4]
+4. Invoke the `logout` method:
+
+[source,java]
+osgiKillbillAPI.getSecurityApi().logout();
```

Comment:
Perhaps worth mentioning this is typically done in a `finally` clause

Reply:
Fixed